test(e2e): add E2E test infra (#561)
oras-project/oras: test/e2e/README.md

> # ORAS End-to-End Testing Dev Guide

Reviewer: Let's update this readme file as it is out of sync.

# ORAS End-to-End Testing Dev Guide

## Setting up
Minimal setup: Run the script in **step 3**

### 1. Common dev setup for ORAS CLI

> https://hackmd.io/_nRHGW8WRfOOvngWc6u-sQ

Reviewer: Please check-in this document in this PR.

Author: Since the original doc invloves vscode-related setup steps, will rewrite this step

### 2. [Optional] Install Ginkgo
This will enable you use `ginkgo` directly in CLI.

> ```

Reviewer: ```suggestion
```shell
```

    ghcr.io/oras-project/registry:latest
```
If the image cannot be pulled, try create a Github PAT and docker/oras login.

### 4. [Optional] Setup backend
```bash
export ORAS_REGISTRY_HOST="localhost:$PORT" # replace with right os/arch
# for PowerShell, use $env:ORAS_REGISTRY_HOST = "localhost:$PORT"
```
If you skip step 4, E2E test will look for distribution ran in `localhost:5000`

### 5. [Optional] Setup ORAS binary for testing
```bash
# Set REPO_ROOT as root folder of oras CLI code
cd $REPO_ROOT
make build
```
### 6. [Optional] Setup pre-built binary
You need to setup below environmental variables to debug a pre-built ORAS binary:
```bash
export ORAS_PATH="bin/linux/amd64/oras" # change target platform if needed
export GITHUB_WORKSPACE=$REPO_ROOT
```
If you skip step 5 or 6, Gomega will build a temp binary, which will include all the CLI code changes in the working directory.

### 7. [Optional] Mount test data
If you want to run command suite, you need to unzip the test file tarball and mount to the distribution
```bash
tar -xvf $REPO_ROOT/test/e2e/testdata/distribution/mount.tar -C $REPO_ROOT/test/e2e/testdata/distribution/

> REPO_ROOT=$(pwd)

Reviewer: What does `pwd` return?

Author: Removing this line and adding more description.

PORT=5000
docker run -dp $PORT:5000 --rm --name oras-e2e \
    --mount type=bind,source=$REPO_ROOT/test/e2e/testdata/distribution/config-example-with-extensions.yml,target=/etc/docker/registry/config.yml \
    --mount type=bind,source=$REPO_ROOT/test/e2e/testdata/distribution/passwd_bcrypt,target=/etc/docker/registry/passwd \

>     --mount type=bind,source=$REPO_ROOT/test/e2e/testdata/distribution/docker,target=/opt/data/registry-root-dir/docker \ # mount test data

Reviewer: Will the comment block fail the bash script?
```suggestion
    --mount type=bind,source=$REPO_ROOT/test/e2e/testdata/distribution/docker,target=/opt/data/registry-root-dir/docker \
```

Author: removed

If you skip step 5 or 6, Gomega will build a temp binary, which will include all the CLI code changes in the working directory.

### 7. [Optional] Mount Test Data

> If you want to run command suite, you need to unzip the zipped data files and mount to the distribution. `REPO_ROOT` points to the root folder of cloned oras CLI code.

Reviewer: ```suggestion
If you want to run command suite, you need to decompress the compressed data files and mount to the distribution. `REPO_ROOT` points to the root folder of cloned oras CLI code.
```
